### DISCUSSION TITLE
Unable to define Schema with Unique Field and Multiple Null Values in Prisma with MongoDB

### DISCUSSION THREAD

**Discussion by innovatorved** [prisma/prisma#18439] [Q&A] [answered]
I'm currently using Prisma ORM with MongoDB and facing an issue with setting a schema that has a unique field with multiple null values. Here's my user model:
```
model User {
  id              String                    @id @default(auto()) @map("_id") @db.ObjectId
  name            String?
  email           String?                   @unique
  username        String?                   @unique
  conversations   ConversationParticipant[]
  messages        Message[]
  emailVerified   DateTime?
  image           String?
  accounts        Account[]
  sessions        Session[]
}
```

I'm having trouble setting the **username** field to allow multiple null values while still maintaining the `@unique` constraint. Prisma only supports one null value in the **username** field, and any attempt to create another user with a null **username** field results in an error.

Note  : I am using next-auth for authentication then create a custom page for username update

I'm currently using Prisma client `@prisma/client": "^4.11.0"`, and `"prisma": "^4.11.0"`.

Look at the Error
```
https://next-auth.js.org/errors#adapter_error_createuser
Invalid `p.user.create()` invocation in
/------/----/-----/frontend/node_modules/@next-auth/prisma-adapter/dist/index.js:6:38

  3 exports.PrismaAdapter = void 0;
  4 function PrismaAdapter(p) {
  5     return {
→ 6         createUser: (data) => p.user.create(
Unique constraint failed on the constraint: `User_username_key` {
  message: '\n' +
    'Invalid `p.user.create()` invocation in\n' +
    '/------/----/-----/frontend/node_modules/@next-auth/prisma-adapter/dist/index.js:6:38\n' +
    '\n' +
    '  3 exports.PrismaAdapter = void 0;\n' +
    '  4 function PrismaAdapter(p) {\n' +
    '  5     return {\n' +
    '→ 6         createUser: (data) => p.user.create(\n' +
    'Unique constraint failed on the constraint: `User_username_key`',
  stack: 'Error: \n' +
    'Invalid `p.user.create()` invocation in\n' +
    '/------/----/-----/frontend/node_modules/@next-auth/prisma-adapter/dist/index.js:6:38\n' +
    '\n' +
    '  3 exports.PrismaAdapter = void 0;\n' +
    '  4 function PrismaAdapter(p) {\n' +
    '  5     return {\n' +
    '→ 6         createUser: (data) => p.user.create(\n' +
    'Unique constraint failed on the constraint: `User_username_key`\n' +
    '    at Zr.handleRequestError (/------/----/-----/frontend/node_modules/@prisma/client/runtime/library.js:171:6414)\n' +
    '    at Zr.handleAndLogRequestError (/------/----/-----/frontend/node_modules/@prisma/client/runtime/library.js:171:5948)\n' +
    '    at Zr.request (/------/----/-----/frontend/node_modules/@prisma/client/runtime/library.js:171:5786)\n' +
    '    at async t._request (/------/----/-----/frontend/node_modules/@prisma/client/runtime/library.js:174:10455)',
  name: 'Error'
}
```

Has anyone else encountered this issue or found a workaround for it? Any suggestions or advice would be greatly appreciated.

**Comment by nurul3101** [comment, marked as answer]
Hello @innovatorved 👋 

There's a related GitHub Issue regarding the same issue that you are running into:
- https://github.com/prisma/prisma/issues/12827
Can you have a look at this issue?

As of now, it's not natively supported in Prisma, but you would need to create a [partial unique index](https://www.mongodb.com/docs/manual/core/index-partial/#std-label-partial-index-with-unique-constraints) directly through MongoDB Atlas.

**Reply by ludralph** [reply]
To add to what @nurul3101  has said MongoDB’s semantics of null is not the same as the semantics of null in SQL databases. As a consequence, unique indexes in MongoDB will raise uniqueness violation error when more than one document has null for the field that has unique index. You're facing an issue because MongoDB treats null values differently than other databases. In MongoDB, null values are treated as unique, and a unique index cannot have more than one document with a null value. This is why you're getting an error when you try to create another user with a null username field.

To work around this issue, you can create a sparse unique index or a partial unique index on the username field that excludes null values but Prisma at the moment does not support sparse unique indexes as can be seen in this open [issue](https://github.com/prisma/prisma/issues/3419)

**Reply by innovatorved** [reply]
Thanks @nurul3101 @ludralph 
But i think its an serious issue because lots of use case rely on this feature. Anyone is working on this ?

**Reply by ludralph** [reply]
It's not prioritized at the moment, but you can leave a 👍 on the open issue to help our engineers prioritize it.

**Reply by innovatorved** [reply]
@ludralph @nurul3101 
I tried to declare direct index but i donot why the error is continous comming. Please modify my schema so i can easily understand

My current schema
```
model User {
  id              String                    @id @default(auto()) @map("_id") @db.ObjectId
  name            String?
  email           String?                   @unique
  username        String?                   @unique
  conversations   ConversationParticipant[]
  messages        Message[]
  emailVerified   DateTime?
  image           String?
  accounts        Account[]
  sessions        Session[]
}
```
DATABASE_NAME=messenger

Please write me the required mogosh query for creating sparse index

**Reply by ludralph** [reply]
The query to create a sparse index is 
```js
db.collection.createIndex({ field: 1 }, { unique: true, sparse: true })
```
Replace `collection` with the name of your collection and `field` with the field you want to create the unique index on.

**Reply by innovatorved** [reply]
Thanks @ludralph 
But i want to ask that should i  have to remove @unique from my username field in Prisma Schema

**Reply by innovatorved** [reply]
Problem resolved Thankyou @ludralph
